### DISCUSSION TITLE
Unable to setup with GPU/CUDA

### DISCUSSION THREAD

**Discussion by ad-mos** [ria-com/nomeroff-net#255] [Q&A] [answered]
I have successfully setup all drivers, CUDA toolkit, cuDNN, builded up openCV with CUDA support, but it's still looks like it ignores system GPU.

I am tested that by running `/examples/py/rest_examples/simple_flask/server.py` with `threaded=True` web serving.
Added some debug output for checking GPU/CUDA status for `cv2` and `pytorch`
![image](https://user-images.githubusercontent.com/22893424/213181029-2b47d7b5-775a-4579-8e9d-eb94803e2953.png)
![image](https://user-images.githubusercontent.com/22893424/213181584-fa51a90d-93b0-4474-b383-a6ad7b8629d3.png)

Performance tested with `ab` benchmark by sending `/detect` requests to `flask`
![image](https://user-images.githubusercontent.com/22893424/213182172-a2f0b8d1-b8e1-441e-80be-cb7c96bb1d4e.png)

CPU usage is 100%, but GPU only 4-12%.
Benchmark shows average proccessing time 1.4s.
![image](https://user-images.githubusercontent.com/22893424/213184028-a6a54e58-f928-43cb-8e19-838019775114.png)

Tried installing different driver versions, cuda toolkits and etc. Wasted few days but no success.
Does anybody know, what am I doing wrong?
Thanks.

**Comment by ad-mos** [comment]
Tried even running docker GPU image (`/docker/build-gpu.sh` + `/docker/run-gpu.sh`) but it can't start even the basic examples and also seems like it's not using GPU.
![image](https://user-images.githubusercontent.com/22893424/213385175-758dd6dc-68d3-426a-a718-957c30177df8.png)

**Comment by ApelSYN** [comment, marked as answer]
We connect YOLO with "default" settings and then change them. In previous versions of yolo you could specify 'cuda', now you have to write the GPU number, for example '0'
You can change the settings for your environment in this file:
[pipes/number_plate_localizators/yolo_v5_detector.py](https://github.com/ria-com/nomeroff-net/blob/master/nomeroff_net/pipes/number_plate_localizators/yolo_v5_detector.py)

`python
model = torch.hub.load(repo_path, 'custom', device="0", path=weights, source="local")`

We will improve it in future versions.

**Reply by ad-mos** [reply]
Thanks, now it shows GPU, but the performance still the same, as a GPU / CPU usage levels = 4-14% / 100%

![image](https://user-images.githubusercontent.com/22893424/213460013-f25d71de-4983-4b46-8a40-6946f5473acf.png)
![image](https://user-images.githubusercontent.com/22893424/213460154-30890d08-0f91-43d8-8537-2e9cf67ebb23.png)
![image](https://user-images.githubusercontent.com/22893424/213460044-581186fd-e87b-49d0-9cee-3c838846bbc6.png)

Is that performance ok? 
P.S. I am using master branch

**Reply by ad-mos** [reply]
Also tested with different images:

1) Big one - 6MB (3472x4624px) + concurency 10
GPU Load = 4-15%
Time for processing 1 image = 1.4s

2) Big one - 6MB (3472x4624px) + concurency 1
GPU Load = 0-5%
Time for processing 1 image = 0.92s

3) Small compressed - 22.6KB (685x514px) + concurency 10
GPU Load = 17-21%
Time for processing 1 image = 0.438s

4) Small compressed - 22.6KB (685x514px) + concurency 1
GPU Load = 12-13%
Time for processing 1 image = 0.063s

**Reply by ad-mos** [reply]
Benchmark from examples:
![image](https://user-images.githubusercontent.com/22893424/213510866-d3f8f853-f30b-4483-939c-6a2cd57e6067.png)

**Reply by ApelSYN** [reply]
Currently, the biggest load is created by **opencv**, which uses the CPU to download photos from disk.

To speed up photo loading, you should use the [libjpeg-turbo-official library](https://libjpeg-turbo.org/).

In the **runtime-test.ipynb** example, the following changes should be made
```
number_plate_detection_and_reading = pipeline("number_plate_detection_and_reading_runtime", 
                                              image_loader="turbo", # Try 'turbo' for faster performance.
                                              quality_profile=[3, 1, 0])
```

You can see in which mode pytorch works with models using this code:
```
print("NUMBERPLATE CLASSIFICATION")
print(number_plate_detection_and_reading.number_plate_classification.detector.model.device)

print("TEXT DETECORS")
for text_detector in number_plate_detection_and_reading.number_plate_text_reading.detector.detectors:
    print(text_detector.model.device)
    
print("YOLO")
print(next(number_plate_detection_and_reading.number_plate_localization.detector.model.parameters()).device)

print("CRAFT")
print(next(number_plate_detection_and_reading.number_plate_key_points_detection.detector.net.parameters()).device)
print(next(number_plate_detection_and_reading.number_plate_key_points_detection.detector.refine_net.parameters()).device)
```
Here is the output from my jupyter notebook:
```
NUMBERPLATE CLASSIFICATION
cuda:0
TEXT DETECORS
cuda:0
cuda:0
cuda:0
cuda:0
cuda:0
cuda:0
cuda:0
cuda:0
cuda:0
cuda:0
YOLO
cuda:0
CRAFT
cuda:0
cuda:0
```

**Reply by ad-mos** [reply]
Debug output of your code:
```
NUMBERPLATE CLASSIFICATION
cuda:0
TEXT DETECORS
cuda:0
cuda:0
cuda:0
cuda:0
cuda:0
cuda:0
cuda:0
cuda:0
cuda:0
cuda:0
YOLO
cuda:0
CRAFT
cuda:0
cuda:0
```

Performance comparison:
1) Big one - 6MB (3472x4624px) + concurency 10
GPU Load = **14-22%** (before 4-15%)
Time for processing 1 image = **0.63s** (before 1.4s)

2) Big one - 6MB (3472x4624px) + concurency 1
GPU Load = **2-6%** (before 0-5%)
Time for processing 1 image = **0.2s** (before 0.92s)

3) Small compressed - 22.6KB (685x514px) + concurency 10
GPU Load = **15-22%** (before 17-21%)
Time for processing 1 image = **0.471s** (before 0.438s)

4) Small compressed - 22.6KB (685x514px) + concurency 1
GPU Load = **12-14%** (before 12-13%)
Time for processing 1 image = **0.058s** (before 0.063s)

Screenshot of benchmarking (Big one - 6MB + concurency 10)
![image](https://user-images.githubusercontent.com/22893424/213730459-e42b0e05-e777-4b1c-b443-653f3ca9e938.png)

**Reply by ad-mos** [reply]
Benchmark timer stats:

**Big one - 6MB (3472x4624px) - one shot**
```
Processed 1 photo
One photo process 0.19008541107177734 seconds

detect_bbox_time_all 0.020519256591796875 per one photo
craft_time_all 0.03821420669555664 per one photo
classification_time_all 0.005549192428588867 per one photo
ocr_time_all 0.009549140930175781 per one photo
127.0.0.1 - - [23/Jan/2023 05:43:07] "POST /detect HTTP/1.1" 200 -
```

**Small compressed - 22.6KB (685x514px) - one shot**
```
Processed 1 photo
One photo process 0.05509495735168457 seconds

detect_bbox_time_all 0.020476818084716797 per one photo
craft_time_all 0.016835451126098633 per one photo
classification_time_all 0.005713224411010742 per one photo
ocr_time_all 0.009476423263549805 per one photo
127.0.0.1 - - [23/Jan/2023 05:43:27] "POST /detect HTTP/1.1" 200 -
```

It's still strange that it can't use 100% of GPU but only 15-25%, but maybe it's okay.

Thank you very much for your help!

**Reply by ad-mos** [reply]
Noticed that replacing **opencv** with **turbo** - greatly reduced recognition accuracy, so unfortunately, it's not usable in most cases for me

**Reply by ApelSYN** [reply]
Show examples on which this problem can be reproduced.

**Reply by ad-mos** [reply]
I will send it on your email

**Comment by ApelSYN** [comment]
This is taking too long to do, I'll go over the settings tomorrow.
